pkg/dnscache/dnscache.go

```diff
@@ -0,0 +1,248 @@
+package dnscache
+
+import (
+	"errors"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru/v2/expirable"
+	"golang.org/x/exp/slices"
+
+	"github.com/aquasecurity/tracee/pkg/logger"
+	"github.com/aquasecurity/tracee/pkg/utils/set"
+	"github.com/aquasecurity/tracee/types/detect"
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+var (
+	ErrDnsCacheIsNotInitialized = errors.New("dns cache is not initialized")
+)
+
+const (
+	DefaultCacheSize    = 500
+	DefaultCacheTTL     = 60 * 5 // 5 minutes in seconds
+	reverseSearchSuffix = ".in-addr.arpa"
+)
+
+type dnsRecords struct {
+	addresses *set.Set[string] // meant for A and AAAA records
+	cname     string
+	alias     string
+}
+
+type DNSCache struct {
+	ipToDNS      *expirable.LRU[string, *set.Set[string]]
+	dnsToRecords *expirable.LRU[string, dnsRecords]
+
+	rwLock *sync.RWMutex
+}
+
+type Config struct {
+	Enable    bool
+	CacheSize int
+	TTL       int
+}
+
+func New(cfg Config) *DNSCache {
+	ttl := time.Second * time.Duration(cfg.TTL)
+	return &DNSCache{
+		ipToDNS:      expirable.NewLRU[string, *set.Set[string]](cfg.CacheSize, nil, ttl),
+		dnsToRecords: expirable.NewLRU[string, dnsRecords](cfg.CacheSize, nil, ttl),
+		rwLock:       &sync.RWMutex{},
+	}
+}
+
+// Add adds the DNS response into the DNS' LRU cache.
+func (dc *DNSCache) Add(dnsResponse trace.DnsResponseData) error {
+	switch dnsResponse.QueryData.QueryType {
+	case "A":
+	case "AAAA":
+	case "PTR":
+	default:
+		return nil // Do not handle other query types
+	}
+
+	dc.rwLock.Lock()
+	defer dc.rwLock.Unlock()
+
+	query := dnsResponse.QueryData.Query
+	answers := extractAnswers(dnsResponse.DnsAnswer)
+	if isReverseQuery(dnsResponse.QueryData.QueryType) {
+		return dc.addReverseQuery(query, answers)
+	}
+
+	// Check if CNAME entry exists.
+	for i, answer := range answers {
+		switch strings.ToUpper(dnsResponse.DnsAnswer[i].Type) {
+		case "CNAME":
+			if err := dc.addDnsCname(query, answer); err != nil {
+				logger.Errorw("Error adding DNS CName record to DNS Cache", "error", err)
+			}
+			goto end
+		case "ALIAS":
+			if err := dc.addDnsAlias(query, answer); err != nil {
+				logger.Errorw("Error adding DNS CName record to DNS Cache", "error", err)
+			}
+			goto end
+		}
+	}
+
+end:
+	return dc.addDnsQuery(query, answers)
+}
+
+// GetIPsForDNS returns IPs by passed DNS from the LRU cache.
+func (dc *DNSCache) GetIPsForDNS(dns string) ([]string, error) {
+	dc.rwLock.RLock()
+	defer dc.rwLock.RUnlock()
+
+	ipList, ok := dc.dnsToRecords.Get(dns)
+	if !ok {
+		return nil, detect.ErrDataNotFound
```

Comment:
I dont think you reuse errors from another package (you're only importing that package because of those AND its weird to use another packages error, IMO).

Reply:
Yeah, there's a leaky abstraction here... I just didn't want to create an error value just so I can translate it again in the data source layer, but this isn't good practice either...

Reply:
Fixed.